--- a/Net/LDAP3.php	2015-03-27 14:59:23.000000000 +0100
+++ b/Net/LDAP3.php	2015-06-30 19:20:40.693675606 +0200
@@ -3060,25 +3060,35 @@
                 $result = false;
             }
         }
-        else {
-            $domain_base_dn = $this->config_get('domain_base_dn');
+        else if ($domain_base_dn = $this->config_get('domain_base_dn')) {
             $domain_filter  = $this->config_get('domain_filter');
-            $name_attribute = $this->config_get('domain_name_attribute');
 
-            if (empty($name_attribute)) {
-                $name_attribute = 'associateddomain';
+            if (strpos($domain_filter, '%s') !== false) {
+                $domain_filter = str_replace('%s', self::quote_string($domain), $domain_filter);
             }
+            else {
+                $name_attribute = $this->config_get('domain_name_attribute');
+                if (empty($name_attribute)) {
+                    $name_attribute = 'associateddomain';
+                }
 
-            $domain_filter = "(&" . $domain_filter . "(" . $name_attribute . "=" . self::quote_string($domain) . "))";
+                $domain_filter = "(&" . $domain_filter . "(" . $name_attribute . "=" . self::quote_string($domain) . "))";
+            }
 
             if ($result = $this->search($domain_base_dn, $domain_filter, 'sub', $attributes)) {
                 $result       = $result->entries(true);
                 $domain_dn    = key($result);
-                $result       = $result[$domain_dn];
-                $result['dn'] = $domain_dn;
 
-                // cache domain DN
-                $this->set_cache_data($ckey, $domain_dn);
+                if (empty($domain_dn)) {
+                    $result = false;
+                }
+                else {
+                    $result       = $result[$domain_dn];
+                    $result['dn'] = $domain_dn;
+
+                    // cache domain DN
+                    $this->set_cache_data($ckey, $domain_dn);
+                }
             }
         }
 
